### PR TITLE
Update smoldot to `0.14.0` and smoldot-light to `0.12.0`

### PR DESCRIPTION
**Pull request by lexnv**
This PR bumps the light client dependencies.
- adjusts the `PlatformRef` implementation
- removes the `with_buffers` module from subxt and relies on the smoldot ones

While at it, bump `which` build dependency to 5.0.0

### CODE REVIEW

File: lightclient/src/platform/wasm_helpers.rs [hunk 1]
```diff
@@ -32,379 +32,11 @@ pub fn sleep(duration: Duration) -> Delay {
...
-                    && !*self.write_closed)
-            {
-                *self.read_write_wake_up_after = Some(self.read_write.now);
-            }
-        }
-    }
-}
+pub struct Stream(
+    #[pin]
+    pub  smoldot::libp2p::with_buffers::WithBuffers<
```
Comment (niklasad1) [inline lightclient/src/platform/wasm_helpers.rs:37]:
```suggestion
    pub smoldot::libp2p::with_buffers::WithBuffers<
```

File: lightclient/src/platform/wasm_platform.rs [hunk 2]
```diff
@@ -142,17 +141,15 @@ impl PlatformRef for SubxtPlatform {
...
-            let socket = WasmSocket::new(addr.as_str()).map_err(|err| ConnectError {
-                message: format!("Failed to reach peer: {err}"),
-            })?;
-
-            Ok(super::wasm_helpers::Stream(
-                super::wasm_helpers::with_buffers::WithBuffers::new(socket),
-            ))
-        })
+        future::ready(super::wasm_helpers::Stream(with_buffers::WithBuffers::new(
+            Box::pin(socket_future),
```
Comment (niklasad1) [inline lightclient/src/platform/wasm_platform.rs:151]: a little annoying to box twice i.e, both the future and the stream but yeye

File: lightclient/src/platform/wasm_helpers.rs
Comment (niklasad1) [inline lightclient/src/platform/wasm_helpers.rs:1]: awesome :tada:

**Review by niklasad1 — APPROVED** [review]

File: lightclient/src/platform/wasm_helpers.rs (hunk 1 above)
Comment (lexnv) [reply lightclient/src/platform/wasm_helpers.rs:37]: I also found this annoying, however it seems this is how rustfmt is formatting this for me 🤔

Comment (jsdw) [reply lightclient/src/platform/wasm_helpers.rs:37]: That's very strange; I haven't seen such double-space formatting before; moving to single space casues a fmt error?

Maybe the macro leads to something weird hmmm

File: lightclient/src/platform/wasm_platform.rs (hunk 2 above)
Comment (jsdw) [reply lightclient/src/platform/wasm_platform.rs:151]: It looks like only one box to me on the future, and the stream isn't boxed? Maybe I haven't dug deeply enough!

File: lightclient/src/platform/wasm_helpers.rs
Comment (jsdw) [reply lightclient/src/platform/wasm_helpers.rs:1]: Yup, awesome to see this stuff going away!

**Review by jsdw — APPROVED** [review]

File: lightclient/src/platform/wasm_helpers.rs (hunk 1 above)
Comment (lexnv) [reply lightclient/src/platform/wasm_helpers.rs:37]: Yep that's super weird:

```bash
cargo fmt --all -- --check                                                              19:42:04
Diff in /home/lexnv/workspace/subxt/lightclient/src/platform/wasm_helpers.rs at line 34:
 #[pin_project::pin_project]
 pub struct Stream(
     #[pin]
-    pub smoldot::libp2p::with_buffers::WithBuffers<
+    pub  smoldot::libp2p::with_buffers::WithBuffers<
         future::BoxFuture<'static, Result<WasmSocket, std::io::Error>>,
         WasmSocket,
         Instant,
```

And also the CI fmt check is passing: https://github.com/paritytech/subxt/actions/runs/7103856224/job/19337498624?pr=1307

**Pull request merged**